insights/parsers/virsh_list_all.py

```diff
@@ -0,0 +1,138 @@
+        True
+        >>> output.get_vm_state('rhel7.0') == 'paused'
+        True
+        >>> output.get_vm_state('rhel9.0') is None
+        True
+        >>> 'cfme' in output
+        False
+        >>> 'cfme-5.7.13' in output
+        True
+
+    Attributes:
+        fields (list): List of ``KeyValue`` namedtupules for each line
+                       in the command.
+
+        cols (list): List id key value pair derived from the command.
+
+        keywords (list): keywords present in the command, each
+                        keyword is converted to lowercase.
+
+    """
+    keyvalue = namedtuple('KeyValue',
+                          ['name', 'state', 'id', 'name_lower'])
+    """namedtuple: Represent name value pair as a namedtuple with case."""
+    def _cleanup(self):
+        for col in self.cols:
+            if col['id'] == '-':
+                col['id'] = None
+            else:
+                col['id'] = (lambda x: int(x) if x.isdigit() else x)(col['id'])
+
+    def parse_content(self, content):
+        self.fields = []
+        self.cols = []
+        self.keywords = []
+        if not content:
+            return
+
+        self.cols = parse_fixed_table(content,
+                                      heading_ignore=['Id', 'Name', 'State'],
+                                      header_substitute=[('Id', 'id'), ('Name', 'name'), ('State', 'state')])[1:]  # noqa
+        self._cleanup()
+
+        for item in self.cols:
+            self.fields.append(self.keyvalue(item['name'], item['state'], item['id'], item['name'].lower()))  # noqa
+        self.keywords = [name.name_lower for name in self.fields]
+
+    def __contains__(self, keyword):
+        return keyword.lower() in self.keywords
+
+    def __iter__(self):
+        return iter(self.fields)
+
+    def search(self, **kw):
+        '''Search item based on key value pair.
+
+        Example:
+
+            >>> len(output.search(state='shut off')) == 11
+            True
+            >>> len(output.search(id=None)) == 11
+            True
+            >>> len(output.search(id=2)) == 1
+            True
+        '''
+        return keyword_search(self.cols, **kw)
+
+    def get_vm_state(self, vmname):
+        '''Get VM state associated with vmname
+
+        Typical output is ``virsh --readonly list --all`` command::
+
+             Id    Name                           State
+            ----------------------------------------------------
+             2     rhel7.4                        running
+             4     rhel7.0                        paused
+
+
+        Example:
+
+            >>> output.get_vm_state('rhel7.0')
+            'paused'
+
+        Args:
+
+            vmname (str): A key. For ex. ``rhel7.0``.
+
+        Returns:
+
+            str: State of VM. Returns None if, ``vmname`` does not exist.
+        '''
+        vmname = vmname.lower()
+        if vmname in self.keywords:
```

Comment:
Though the code looks simple, don't you think `search` is too heavy? how about implement it by iterating the `self.cols` like:

```
st = [i['state'] for i in self.cols if i['name'] == vmname]
return st[0] if st else None
```